apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -274,6 +278,40 @@ const (
 	PreemptionPolicyLowerOrNewerEqualPriority PreemptionPolicy = "LowerOrNewerEqualPriority"
 )
 
+type FlavorFungibilityPolicy string
+
+const (
+	Borrow        FlavorFungibilityPolicy = "Borrow"
+	Preempt       FlavorFungibilityPolicy = "Preempt"
+	TryNextFlavor FlavorFungibilityPolicy = "TryNextFlavor"
+)
+
+// FlavorFungibility determines whether a workload should try the next flavor
+// before borrowing or preempting in current flavor.
+type FlavorFungibility struct {
+	// whenCanBorrow determines whether a workload should try the next flavor
+	// before borrowing in current flavor. The possible values are:
+	//
+	// - `Borrow` (default): allocate in current flavor if borrowing
+	//   is possible.
+	// - `TryNextFlavor`: try next flavor even if the current
+	//   flavor has enough resources to borrow.
+	//
+	// +kubebuilder:validation:Enum={Borrow,TryNextFlavor}
+	// +kubebuilder:default="Borrow"
+	WhenCanBorrow FlavorFungibilityPolicy `json:"whenCanBorrow,omitempty"`
+	// whenCanPreempt determines whether a workload should try the next flavor
+	// before borrowing in current flavor. The possible values are:
+	//
+	// - `Preempt`: allocate in current flavor if it's possible to preempt some workloads.
+	// - `TryNextFlavor` (default): try next flavor even if there are enough
+	//   candidates for preemption in the current flavor.
+	//
+	// +kubebuilder:validation:Enum={Preempt,TryNextFlavor}
+	// +kubebuilder:default="TryNextFlavor"
```

Comment:
When `whenCanPreempt=Preempt`, it means that we will pause at each flavor where usage+request<nominal+borrowingLimit and attempt to preempt. When `whenCanPreempt=tryNextFlavor`, it means that we will try to preempt after traversing all flavors and when mode=Preempt. `whenCanPreempt=tryNextFlavor` is the current default behavior because we will traverse all flavors and break the loop until the best mode is `Fit` or all flavors have been .

Reply:
Ah, right. The flavorassigner traverses all flavors if the mode is `Preempt`.
Thank you for the clarifications, and Sorry for the confusion.

https://github.com/kubernetes-sigs/kueue/blob/c4197436d17e6241cd71d880c68d4329fa2f65c8/pkg/scheduler/flavorassigner/flavorassigner.go#L392-L399